File: jvm/trino-ducklake/src/dev/brikk/ducklake/trino/plugin/InProcessDuckDbExecutor.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.brikk.ducklake.trino.plugin;

import io.airlift.log.Logger;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.ipc.ArrowReader;
import org.duckdb.DuckDBConnection;
import org.duckdb.DuckDBResultSet;

import java.io.IOException;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Optional;

import static java.lang.String.format;

/**
 * In-process {@link DucklakeDuckDbExecutor}: per-split embedded DuckDB via
 * {@code jdbc:duckdb:}. The {@code .db} file is ATTACHed either as a local
 * path (materialize cache) or as an {@code s3://} URL via the httpfs extension,
 * exactly as the original {@code DuckDbFilePageSource} did before the executor
 * abstraction was introduced. Behaviour-preserving: nothing about query
 * execution changes vs the pre-refactor code.
 */
final class InProcessDuckDbExecutor
        implements DucklakeDuckDbExecutor
{
    private static final Logger log = Logger.get(InProcessDuckDbExecutor.class);

    private static final long ARROW_BATCH_SIZE = 1024;
    private static final String ATTACHED_DB = "ducklake_in";
    private static final String ATTACHED_TABLE = "t";

    @Override
    public ExecutionContext execute(ExecutionRequest request)
            throws SQLException
    {
        DuckDBConnection connection = (DuckDBConnection) DriverManager.getConnection("jdbc:duckdb:");
        Statement statement = null;
        DuckDBResultSet resultSet = null;
        BufferAllocator allocator = null;
        ArrowReader arrowReader = null;
        try {
            try (Statement attachStmt = connection.createStatement()) {
                attachStmt.execute(buildAttachSql(request.target(), attachStmt));
            }
            String selectSql = buildSelectSql(request);
            statement = connection.createStatement();
            resultSet = (DuckDBResultSet) statement.executeQuery(selectSql);
            allocator = new RootAllocator();
            arrowReader = (ArrowReader) resultSet.arrowExportStream(allocator, ARROW_BATCH_SIZE);
            return new InProcessExecutionContext(connection, statement, resultSet, allocator, arrowReader,
                    describeAttachTarget(request.target()));
        }
        catch (SQLException | RuntimeException e) {
            // Surface the most-likely-useful error message; close half-built resources.
            closeQuietly(arrowReader, resultSet, statement, allocator);
            try {
                connection.close();
            }
            catch (SQLException ignored) {
            }
            throw e;
        }
    }

    private static String buildAttachSql(DuckDbAttachTarget target, Statement stmtForSideEffects)
            throws SQLException
    {
        return switch (target) {
            case DuckDbAttachTarget.LocalPath local -> format(
                    "ATTACH '%s' AS %s (READ_ONLY)",
                    local.path().toAbsolutePath().toString().replace("'", "''"),
                    ATTACHED_DB);
            case DuckDbAttachTarget.HttpfsS3 remote -> {
                // INSTALL is idempotent and cached on disk per-DuckDB-version. LOAD
                // is required per-instance. CREATE SECRET supplies S3 credentials
                // for the ATTACH that follows; fixed name means re-creation on a
                // fresh in-memory instance always succeeds.
                stmtForSideEffects.execute("INSTALL httpfs");
                stmtForSideEffects.execute("LOAD httpfs");
                stmtForSideEffects.execute(remote.s3Config().renderCreateSecretSql());
                yield format(
                        "ATTACH '%s' AS %s (READ_ONLY)",
                        remote.s3Url().replace("'", "''"),
                        ATTACHED_DB);
            }
        };
    }

    private static String buildSelectSql(ExecutionRequest request)
    {
        StringBuilder sql = new StringBuilder("SELECT ");
        if (request.isEmptyProjection()) {
            // COUNT(*) and similar collapse to no projected columns. We still need a
            // SELECT clause that yields one row per file row so deletes can be
            // applied by row position downstream — emit a constant, ignored by the
            // converter.
            sql.append("1");
        }
        else {
            var columns = request.projectedColumns();
            for (int i = 0; i < columns.size(); i++) {
                if (i > 0) {
                    sql.append(", ");
                }
                String name = columns.get(i).columnName().replace("\"", "\"\"");
                sql.append('"').append(name).append('"');
            }
        }
        sql.append(" FROM ").append(ATTACHED_DB).append(".main.").append(ATTACHED_TABLE);
        Optional<String> whereClause = DuckDbWhereClauseTranslator.toWhereClause(request.pushedPredicate());
        whereClause.ifPresent(w -> sql.append(" WHERE ").append(w));
        return sql.toString();
    }

    private static String describeAttachTarget(DuckDbAttachTarget target)
    {
        return switch (target) {
            case DuckDbAttachTarget.LocalPath local -> local.path().toString();
            case DuckDbAttachTarget.HttpfsS3 remote -> remote.s3Url();
        };
    }

    private static void closeQuietly(AutoCloseable... resources)
    {
        for (AutoCloseable r : resources) {
            if (r == null) {
                continue;
            }
            try {
                r.close();
            }
            catch (Exception ignored) {
            }
        }
    }

    private static final class InProcessExecutionContext
            implements ExecutionContext
    {
        private final DuckDBConnection connection;
        private final Statement statement;
        private final DuckDBResultSet resultSet;
        private final BufferAllocator allocator;
        private final ArrowReader arrowReader;
        private final String attachDescription;

        InProcessExecutionContext(
                DuckDBConnection connection,
                Statement statement,
                DuckDBResultSet resultSet,
                BufferAllocator allocator,
                ArrowReader arrowReader,
                String attachDescription)
        {
            this.connection = connection;
            this.statement = statement;
            this.resultSet = resultSet;
            this.allocator = allocator;
            this.arrowReader = arrowReader;
            this.attachDescription = attachDescription;
        }

        @Override
        public ArrowReader arrowReader()
        {
            return arrowReader;
        }

        @Override
        public long memoryUsage()
        {
            return allocator == null ? 0 : allocator.getAllocatedMemory();
        }

        @Override
        public void close() throws IOException
        {
            Throwable suppressed = null;
            try {
                if (arrowReader != null) {
                    arrowReader.close();
                }
            }
            catch (Throwable t) {
                suppressed = t;
            }
            try {
                if (resultSet != null) {
                    resultSet.close();
                }
            }
            catch (Throwable t) {
                if (suppressed == null) {
                    suppressed = t;
                }
            }
            try {
                if (statement != null) {
                    statement.close();
                }
            }
            catch (Throwable t) {
                if (suppressed == null) {
                    suppressed = t;
                }
            }
            if (connection != null) {
                try (Statement detach = connection.createStatement()) {
                    detach.execute("DETACH " + ATTACHED_DB);
                }
                catch (Throwable t) {
                    if (suppressed == null) {
                        suppressed = t;
                    }
                }
                try {
                    connection.close();
                }
                catch (Throwable t) {
                    if (suppressed == null) {
                        suppressed = t;
                    }
                }
            }
            try {
                if (allocator != null) {
                    allocator.close();
                }
            }
            catch (Throwable t) {
                if (suppressed == null) {
                    suppressed = t;
                }
            }
            if (suppressed != null) {
                log.warn(suppressed, "Error while closing in-process DuckDB executor for %s", attachDescription);
            }
        }
    }
}

File: jvm/trino-ducklake/src/dev/brikk/ducklake/trino/plugin/InProcessDuckDbExecutor.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.brikk.ducklake.trino.plugin

import io.airlift.log.Logger
import org.apache.arrow.memory.BufferAllocator
import org.apache.arrow.memory.RootAllocator
import org.apache.arrow.vector.ipc.ArrowReader
import org.duckdb.DuckDBConnection
import org.duckdb.DuckDBResultSet
import java.io.IOException
import java.lang.String.format
import java.sql.DriverManager
import java.sql.SQLException
import java.sql.Statement
import java.util.Properties
import java.util.concurrent.ConcurrentHashMap

/**
 * In-process [DucklakeDuckDbExecutor]: per-split embedded DuckDB via
 * `jdbc:duckdb:`. The `.db` file is ATTACHed either as a local
 * path (materialize cache) or as an `s3://` URL via the httpfs extension,
 * exactly as the original `DuckDbFilePageSource` did before the executor
 * abstraction was introduced. Behaviour-preserving: nothing about query
 * execution changes vs the pre-refactor code.
 */
class InProcessDuckDbExecutor
internal constructor(private val tuning: DuckDbTuning, private val parityExtensionPath: String) : DucklakeDuckDbExecutor {

    /**
     * Test-only convenience constructor: auto-resolves the bundled trino_parity
     * extension binary from the plugin classpath. Throws if no bundled binary
     * is available for the host platform — production code paths the path
     * through the factory which surfaces a CONFIGURATION_INVALID error instead.
     */
    internal constructor() : this(DuckDbTuning.defaults(), TrinoParityExtensionResolver.resolveBundledExtensionPath()
            ?: throw IllegalStateException(
                    "No bundled trino_parity extension found on classpath. " +
                            "Build the extension first: `(cd duckdb-trino-parity-extension && GEN=ninja make)`, " +
                            "then rebuild the trino-ducklake plugin jar."))

    @Throws(SQLException::class)
    override fun execute(request: DucklakeDuckDbExecutor.ExecutionRequest): DucklakeDuckDbExecutor.ExecutionContext {
        // allow_unsigned_extensions is a database-startup setting; setting it via
        // `SET` after the DB is running fails ("Cannot change while running") AND
        // closes the JDBC Statement on the way out. Pass it as a connection
        // property so the LOAD '<path>' below works.
        val connectionProps = Properties()
        connectionProps.setProperty("allow_unsigned_extensions", "true")
        val connection: DuckDBConnection = DriverManager.getConnection("jdbc:duckdb:", connectionProps) as DuckDBConnection
        var statement: Statement? = null
        var resultSet: DuckDBResultSet? = null
        var allocator: BufferAllocator? = null
        var arrowReader: ArrowReader? = null
        try {
            TrinoFunctionAliases.loadInProcess(connection, parityExtensionPath)
            val sourceRef: String = connection.createStatement().use { attachStmt ->
                DuckDbTuningSql.applyDirect(attachStmt, tuning)
                val ref: String = prepareSource(request.target(), attachStmt)
                applySessionTimeZone(attachStmt, request.duckDbTimeZone())
                ref
            }
            val selectSql = DuckDbSelectSqlBuilder.buildSelectSql(sourceRef, request)
            statement = connection.createStatement()
            resultSet = statement.executeQuery(selectSql) as DuckDBResultSet
            allocator = RootAllocator()
            arrowReader = resultSet.arrowExportStream(allocator, ARROW_BATCH_SIZE) as ArrowReader
            return InProcessExecutionContext(connection, statement, resultSet, allocator, arrowReader,
                    describeAttachTarget(request.target()))
        }
        catch (e: SQLException) {
            // Surface the most-likely-useful error message; close half-built resources.
            closeQuietly(arrowReader, resultSet, statement, allocator)
            try {
                connection.close()
            }
            catch (ignored: SQLException) {
            }
            throw e
        }
        catch (e: RuntimeException) {
            // Surface the most-likely-useful error message; close half-built resources.
            closeQuietly(arrowReader, resultSet, statement, allocator)
            try {
                connection.close()
            }
            catch (ignored: SQLException) {
            }
            throw e
        }
    }

    private class InProcessExecutionContext(
            private val connection: DuckDBConnection?,
            private val statement: Statement?,
            private val resultSet: DuckDBResultSet?,
            private val allocator: BufferAllocator?,
            private val arrowReader: ArrowReader?,
            private val attachDescription: String) : DucklakeDuckDbExecutor.ExecutionContext {
        override fun arrowReader(): ArrowReader = arrowReader!!

        override fun memoryUsage(): Long =
            allocator?.allocatedMemory ?: 0

        @Throws(IOException::class)
        override fun close() {
            var suppressed: Throwable? = null
            try {
                arrowReader?.close()
            }
            catch (t: Throwable) {
                suppressed = t
            }
            try {
                resultSet?.close()
            }
            catch (t: Throwable) {
                if (suppressed == null) {
                    suppressed = t
                }
            }
            try {
                statement?.close()
            }
            catch (t: Throwable) {
                if (suppressed == null) {
                    suppressed = t
                }
            }
            if (connection != null) {
                // No explicit DETACH: this is a fresh per-split in-memory `jdbc:duckdb:` connection
                // that is never reused, so connection.close() tears down the attached DB anyway.
                // DETACH-on-close is meaningful only for engines whose server-side ATTACH state
                // persists across queries (see DucklakeDuckDbExecutor.ExecutionContext) — the
                // in-process path is not such an engine, so the extra round trip (and its spurious
                // close-time WARN on an already-tearing-down connection) was pure overhead.
                try {
                    connection.close()
                }
                catch (t: Throwable) {
                    if (suppressed == null) {
                        suppressed = t
                    }
                }
            }
            try {
                allocator?.close()
            }
            catch (t: Throwable) {
                if (suppressed == null) {
                    suppressed = t
                }
            }
            if (suppressed != null) {
                log.warn(suppressed, "Error while closing in-process DuckDB executor for %s", attachDescription)
            }
        }
    }

    companion object {
        private val log: Logger = Logger.get(InProcessDuckDbExecutor::class.java)

        private const val ARROW_BATCH_SIZE: Long = 1024
        private const val ATTACHED_DB: String = "ducklake_in"
        private const val ATTACHED_TABLE: String = "t"

        /**
         * Prepare the per-split DuckDB instance for [target] (run ATTACH or INSTALL/LOAD) and
         * return the SQL FROM source — a `db.main.t` reference for attach targets, or a
         * `scanFn('path')` table-function call for a [DuckDbAttachTarget.FileScan]. The caller
         * passes the result to [DuckDbSelectSqlBuilder.buildSelectSql].
         */
        @Throws(SQLException::class)
        private fun prepareSource(target: DuckDbAttachTarget, stmt: Statement): String =
            when (target) {
                is DuckDbAttachTarget.LocalPath -> {
                    stmt.execute(format(
                            "ATTACH '%s' AS %s (READ_ONLY)",
                            target.path.toAbsolutePath().toString().replace("'", "''"),
                            ATTACHED_DB))
                    "$ATTACHED_DB.main.$ATTACHED_TABLE"
                }
                is DuckDbAttachTarget.HttpfsS3 -> {
                    // INSTALL is idempotent and cached on disk per-DuckDB-version. LOAD
                    // is required per-instance. CREATE SECRET supplies S3 credentials
                    // for the ATTACH that follows; fixed name means re-creation on a
                    // fresh in-memory instance always succeeds.
                    stmt.execute("INSTALL httpfs")
                    stmt.execute("LOAD httpfs")
                    stmt.execute(target.s3Config.renderCreateSecretSql())
                    stmt.execute(format(
                            "ATTACH '%s' AS %s (READ_ONLY)",
                            target.s3Url.replace("'", "''"),
                            ATTACHED_DB))
                    "$ATTACHED_DB.main.$ATTACHED_TABLE"
                }
                is DuckDbAttachTarget.FileScan -> {
                    // No ATTACH: read the single file via the extension's table function.
                    // extension/scanFunction are connector-controlled constants (not user input).
                    stmt.execute("INSTALL ${target.extension}")
                    stmt.execute("LOAD ${target.extension}")
                    if (target.s3Config.isPresent) {
                        stmt.execute("INSTALL httpfs")
                        stmt.execute("LOAD httpfs")
                        stmt.execute(target.s3Config.get().renderCreateSecretSql())
                    }
                    "${target.scanFunction}('${target.path.replace("'", "''")}'${target.extraArgsSql})"
                }
            }

        private fun describeAttachTarget(target: DuckDbAttachTarget): String =
            when (target) {
                is DuckDbAttachTarget.LocalPath -> target.path.toString()
                is DuckDbAttachTarget.HttpfsS3 -> target.s3Url
                is DuckDbAttachTarget.FileScan -> target.path
            }

        private fun closeQuietly(vararg resources: AutoCloseable?) {
            for (r in resources) {
                if (r == null) {
                    continue
                }
                try {
                    r.close()
                }
                catch (ignored: Exception) {
                }
            }
        }

        private val TIMEZONE_FAILURE_WARNED: ConcurrentHashMap<String, Boolean> =
                ConcurrentHashMap()

        /**
         * Best-effort `SET TimeZone = '<zone>'` after attach. Empty Optional
         * is a no-op (test harness paths without a session). On SQLException we log
         * a one-shot WARN per (already-normalised) zone string and proceed — the
         * connection is still usable for Tier A/B work which doesn't care about the
         * session zone. Tier C correctness is compromised for this split, but
         * Tier C functions stay off the pushdown surface until chunk 3 anyway,
         * so the only visible effect today is that `current_setting('TimeZone')`
         * on the executor side reflects the JVM system default rather than Trino's
         * session zone.
         */
        private fun applySessionTimeZone(stmt: Statement, zone: java.util.Optional<String>) {
            if (zone.isEmpty) {
                return
            }
            val z = zone.get()
            // Single-quote escape: defence in depth — the normaliser only emits values
            // from a fixed grammar (IANA zone names, Etc/GMT±N, the original Trino
            // string for fractional offsets), none of which contain quotes, but the
            // string comes from session config and we treat it as untrusted by habit.
            val sql = "SET TimeZone = '" + z.replace("'", "''") + "'"
            try {
                stmt.execute(sql)
            }
            catch (e: SQLException) {
                if (TIMEZONE_FAILURE_WARNED.putIfAbsent(z, true) == null) {
                    log.warn("DuckDB rejected SET TimeZone for normalised zone '%s' (in-process): %s. "
                                    + "Subsequent splits with the same zone proceed without an explicit "
                                    + "SET; Tier A/B pushdown unaffected, Tier C correctness may diverge. "
                                    + "See dev-docs/archive/REPORT-datetime-tz-handling.md.",
                            z, e.message?.lineSequence()?.firstOrNull() ?: e.message ?: "(no message)")
                }
            }
        }
    }
}
